modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsPersistenceStoreImpl.java

```diff
@@ -0,0 +1,472 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query.stat;
+
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.processors.cache.persistence.IgniteCacheDatabaseSharedManager;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.MetastorageLifecycleListener;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.ReadOnlyMetastorage;
+import org.apache.ignite.internal.processors.cache.persistence.metastorage.ReadWriteMetastorage;
+import org.apache.ignite.internal.processors.query.stat.messages.StatsKeyMessage;
+import org.apache.ignite.internal.processors.query.stat.messages.StatsObjectData;
+import org.apache.ignite.internal.processors.subscription.GridInternalSubscriptionProcessor;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Sql statistics storage in metastore.
+ * Will store all statistics related objects with prefix "stats."
+ * Store only partition level statistics.
+ */
+public class IgniteStatisticsPersistenceStoreImpl implements IgniteStatisticsStore, MetastorageLifecycleListener {
+    /** In local meta store it store partitions statistics by path: stats.<SCHEMA>.<OBJECT>.<partId> */
+    private static final String META_SEPARATOR = ".";
+
+    /** Local metastore statistics prefix. */
+    private static final String META_STAT_PREFIX = "stats";
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Database shared manager. */
+    private final IgniteCacheDatabaseSharedManager db;
+
+    /** Statistics repository. */
+    private final IgniteStatisticsRepository repo;
+
+    /** Metastorage. */
+    private volatile ReadWriteMetastorage metastore;
+
+    /**
+     * Constructor.
+     *
+     * @param subscriptionProcessor Grid subscription processor to track metastorage availability.
+     * @param db Database shared manager to lock db while reading/writing metastorage.
+     * @param repo Repository to fulfill on metastore available.
+     * @param logSupplier Logger getting function.
+     */
+    public IgniteStatisticsPersistenceStoreImpl(
+            GridInternalSubscriptionProcessor subscriptionProcessor,
+            IgniteCacheDatabaseSharedManager db,
+            IgniteStatisticsRepository repo,
+            Function<Class<?>, IgniteLogger> logSupplier
+    ) {
+        this.db = db;
+        this.repo = repo;
+        subscriptionProcessor.registerMetastorageListener(this);
```

Comment:
Is it ever unregister?

Reply:
This statistic store either should have lifecycle methods or should be subscribed from outside?

Reply:
GridInternalSubscriptionProcessor events looks like one-time events and there are no methods like unregisterMetastorageListener at all. Neither have documentation on most of methods.